### PR TITLE
[terra-data-grid] Update to automatically focus on single buttons or hyperlinks in cells

### PR DESCRIPTION
**Pull request by sdadn**
### Summary

This PR updates the focus behavior to negate the need for dive-in mode in cells with only a single button or hyperlink and focus on them directly.

### Testing

This change was tested using:

- [x] WDIO
- [ ] Jest
- [ ] Visual testing (please attach a screenshot or recording)
- [ ] Other (please describe below)
- [ ] No tests are needed

### Reviews

In addition to engineering reviews, this PR needs:

- [ ] UX review
- [ ] Accessibility review
- [ ] Functional review

### Additional Details

**This PR resolves:**

UXPLATFORM-10225

---

Thank you for contributing to Terra.
@cerner/terra

### CODE REVIEW

**Review by kenk2 — APPROVED** [review]

File: packages/terra-table/src/utils/focusableButton.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// This is a temporary file for the purpose of adding a new feature without need for refactoring across multiple components.
+// This will be removed is a separate refactor effort.
+
+const getFocusableButtonsOrHyperlinks = (parentElement) => {
+  const focusableElementSelector = 'a[href]:not([tabindex=\'-1\']), button:not([disabled]):not([tabindex=\'-1\'])';
```
Comment (adoroshk) [inline packages/terra-table/src/utils/focusableButton.js:5]: I am wondering if we need to include buttons that are created like <div> tags with a role="button", or that is out of scope?

Comment (sdadn) [reply packages/terra-table/src/utils/focusableButton.js:5]: I believe custom elements are out of scope for this feature.

**Review by adoroshk — APPROVED** [review]

File: packages/terra-framework-docs/src/terra-dev-site/doc/data-grid/WorklistDataGrid.1/Examples.3/WorklistDataGridFocusableCell.doc.mdx [hunk 2]
```diff
@@ -10,10 +10,13 @@ To make these keys available for cell navigation as part of its basic functional
 Enabling single cell navigation will disable grid navigation and disabling single cell navigation will restore grid navigation.
 
 <span style={{color: 'blue'}}>
-<em>Single Cell navigation is built-in and does not require any additional setup or implementation.</em>
+<b>Notes</b>
+<ul>
+  <li><em>Single Cell navigation is built-in and does not require any additional setup or implementation.</em></li>
+  <li><em>Cells which contain only a focusable button or hyperlink will focus on those components directly without needing to diving into the cell.</em></li>
```
Comment (nikhitasharma) [inline packages/terra-framework-docs/src/terra-dev-site/doc/data-grid/WorklistDataGrid.1/Examples.3/WorklistDataGridFocusableCell.doc.mdx:16]:
```suggestion
  <li><em>Cells which contain only a focusable button or hyperlink will focus on those components directly without needing to dive into the cell.</em></li>
```

File: packages/terra-data-grid/tests/wdio/data-grid-spec.js [hunk 3]
```diff
@@ -280,6 +264,26 @@ Terra.describeViewports('DataGrid', ['medium', 'large'], () => {
       Terra.validates.element('data-grid-focusable-select-retains-focus', { columnResizeSelector });
       expect(browser.$('#specialties').isFocused()).toBe(true);
     });
+
+    it('focuses on a button if it is the only component in a cell', () => {
```
Comment (nikhitasharma) [inline packages/terra-data-grid/tests/wdio/data-grid-spec.js:268]: Should we have a test with more than one focusable component in a cell?

File: packages/terra-table/src/subcomponents/Cell.jsx [hunk 4]
```diff
@@ -213,7 +227,15 @@ function Cell(props) {
...
   }, [isGridContext]);
 
-  const onMouseDown = (event) => {
+  const handleFocus = () => {
+    const element = hasSingleButtonOrHyperlinkElement();
+
+    if (element) {
+      element.focus();
+    }
+  };
```
Comment (cm9361) [inline packages/terra-table/src/subcomponents/Cell.jsx:236]: Is this correct?  We don't want this behavior for the Terra Table.  This functionality should only be possible for a DataGrid component.

File: packages/terra-table/src/utils/focusableButton.js (hunk 1 above)
Comment (cm9361) [reply packages/terra-table/src/utils/focusableButton.js:5]: I think you would actually have to account for this.  You might need to exclude certain types rather than being so specific.  This selector seems too restrictive.

File: packages/terra-table/src/subcomponents/Cell.jsx [hunk 5]
```diff
@@ -213,7 +227,15 @@ function Cell(props) {
     }
   }, [isGridContext]);
 
-  const onMouseDown = (event) => {
+  const handleFocus = () => {
```
Comment (cm9361) [inline packages/terra-table/src/subcomponents/Cell.jsx:230]: I don't think we want to add this logic to the onFocus synthetic event.

File: packages/terra-data-grid/tests/wdio/data-grid-spec.js (hunk 3 above)
Comment (sdadn) [reply packages/terra-data-grid/tests/wdio/data-grid-spec.js:268]: I think that should be covered here:

https://github.com/cerner/terra-framework/blob/b4b9f88804cb4484fc16269a650c55c56f299bdb/packages/terra-data-grid/tests/wdio/data-grid-spec.js#L234-L239

File: packages/terra-table/src/subcomponents/Cell.jsx
Comment (cm9361) [inline packages/terra-table/src/subcomponents/Cell.jsx:1]: You did not account for the isInteractable state variable.  The voice reader commands will not be correct in the case that there is a button element based on the current changes.

File: packages/terra-table/src/subcomponents/Cell.jsx [hunk 6]
```diff
@@ -266,7 +266,13 @@ function Cell(props) {
         case KeyCode.KEY_RETURN:
           // Lock focus into component
           if (isGridContext && hasFocusableElements()) {
```
Comment (cm9361) [inline packages/terra-table/src/subcomponents/Cell.jsx:268]: It would probably be better to create a method that checks for elements that are not input based fields.  You need a way to keep your conditions in sync.  Doing it this way won't allow that to happen.

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 7]
```diff
@@ -598,11 +616,11 @@ const DataGrid = forwardRef((props, ref) => {
           columnHeaderHeight={columnHeaderHeight}
           rowHeight={rowHeight}
           rowHeaderIndex={rowHeaderIndex}
+          onCellSelect={onCellSelect}
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:619]: Did we need to move this?  It was previously with the other selection callbacks.

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 8]
```diff
@@ -252,31 +255,46 @@ const DataGrid = forwardRef((props, ref) => {
...
-      } else {
-        // Set focus on input field (checkbox) of row selection cells.
-        focusedCell = grid.current.rows[newRowIndex].cells[newColIndex];
-        if (isRowSelectionCell(newColIndex) && focusedCell.getElementsByTagName('input').length > 0) {
-          [focusedCell] = focusedCell.getElementsByTagName('input');
-        }
+    let focusedCell;
 
-        // Set focus to column header button, if it exists
-        const isHeaderRow = (newRowIndex === 0 || (hasColumnHeaderActions && newRowIndex === 1));
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:271]: What happened to the header row logic?

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 9]
```diff
@@ -252,31 +255,46 @@ const DataGrid = forwardRef((props, ref) => {
...
+
+    // Set focus on input field (checkbox) of row selection cells.
+    if (isRowSelectionCell(newColIndex) && focusedCell.getElementsByTagName('input').length === 1) {
+      [focusedCell] = focusedCell.getElementsByTagName('input');
+      focusedCell?.focus();
+      return;
+    }
+
+    // Set focus to a single header button or hyperlink if they are the only content in cell
+    if (!focusedCell.hasAttribute('tabindex') || focusedCell.getElementsByTagName('button').length === 1 || focusedCell.getElementsByTagName('a').length === 1) {
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:290]: Are we really limiting this to specifically button and anchor tags?  What if I had a div with a role of button, for example?  This logic seems a little too restrictive.

File: packages/terra-data-grid/src/DataGrid.jsx [hunk 10]
```diff
@@ -252,31 +255,46 @@ const DataGrid = forwardRef((props, ref) => {
...
+      focusedCell = focusedCell.querySelector('[role="button"]') || focusedCell.querySelector('button') || focusedCell.querySelector('a');
```
Comment (cm9361) [inline packages/terra-data-grid/src/DataGrid.jsx:291]: If you are using a querySelector, you could use a single selector instead of having 3 separate conditions to query the DOM.  You can use commas in your selector to separate.

File: packages/terra-data-grid/src/DataGrid.jsx (hunk 7 above)
Comment (sdadn) [reply packages/terra-data-grid/src/DataGrid.jsx:619]: This was an abandoned attempt at alphabetization, reverted.

File: packages/terra-data-grid/src/DataGrid.jsx (hunk 9 above)
Comment (sdadn) [reply packages/terra-data-grid/src/DataGrid.jsx:290]: I believe custom components were excluded, which is why that wasn't considered.  Unless if I misunderstood the requirements. @mjpalazzo are divs with button roles included?

File: packages/terra-data-grid/src/DataGrid.jsx (hunk 8 above)
Comment (sdadn) [reply packages/terra-data-grid/src/DataGrid.jsx:271]: Since header rows with action buttons also only have a single button, this overlaps with the current story to add focus to single buttons in any row. The isHeaderRow condition was removed, the rest of the logic was tweaked to accommodate single buttons/hyperlinks in any row.

File: packages/terra-data-grid/src/DataGrid.jsx (hunk 10 above)
Comment (sdadn) [reply packages/terra-data-grid/src/DataGrid.jsx:291]: Updated here: d86a31695501e5e103ba9f36adff2a349fedc074

File: packages/terra-table/src/subcomponents/Cell.jsx (hunk 4 above)
Comment (sdadn) [reply packages/terra-table/src/subcomponents/Cell.jsx]: Moved this to the data grid component.

File: packages/terra-table/src/subcomponents/Cell.jsx (hunk 6 above)
Comment (sdadn) [reply packages/terra-table/src/subcomponents/Cell.jsx:282]: Added.

File: packages/terra-data-grid/src/DataGrid.jsx (hunk 8 above)
Comment (adoroshk) [reply packages/terra-data-grid/src/DataGrid.jsx:273]: I tested the action button behavior and confirm that it works as expected after that change. As action button is a `<button>` element and the behavior for it is now a default behavior, it makes sense that the `isHeaderRow` condition is no longer needed.

**Comment by mjpalazzo** [comment]
@sdadn - Overall this enhancement is working well. I have a couple of minor findings. 

1. When focus is on the hyperlink, the black solid line focus indicator is cut off on the left side. See attached screen shots. Focus indicator on button and cells look fine. See screenshots for these as well.
2. This is a nit but the column header text in the example is incorrect. For example the header that has "Column 1" in the header is actually column 2. This causes the screen reader to announce "Column 1" some helper text and then "Column 2"

WDG interactive cell 3 focus issue
WDG interactive cell 2
WDG interactive cell 1

**Comment by sdadn** [comment]
The WDIO failure is caused by an another component. Terra-data-grid tests are all passing:

CleanShot 2024-03-14 at 09 27 36

CleanShot 2024-03-14 at 09 27 00

CleanShot 2024-03-14 at 09 27 14

**Review by nikhitasharma — APPROVED** [review]

**Review by adoroshk — APPROVED** [review]

File: packages/terra-table/src/subcomponents/Cell.jsx
Comment (sdadn) [reply packages/terra-table/src/subcomponents/Cell.jsx:1]: The voicereader output seems to be correct based on the accessibility review.
